### PR TITLE
feat: configuration setting for running detailed parse

### LINKED ISSUE #1334: Presence of an argument?
Say I have the arguments --abc N. If I make the default for abc be 5, then yargs will create the .abc value with 5. However, this doesn't tell me if --abc was actually seen on the command line. Is there a way to see an argument was given at all?

### PR DESCRIPTION
**Pull request by bcoe**
You can now set `yargs.parserConfiguration({detailed: true})` and receive detailed results from `yargs-parser`.

This provides an interface into @juergba's [defaulted functionality](https://github.com/yargs/yargs-parser/pull/211).

fixes #1334 

CC: @coreyfarrell.

### CODE REVIEW

File: lib/command.js [hunk 1]
```diff
@@ -367,17 +383,30 @@ module.exports = function command (yargs, usage, validation, globalMiddleware) {
...
         if (positionalKeys.indexOf(key) !== -1) {
-          // any new aliases need to be placed in positionalMap, which
-          // is used for validation.
-          if (!positionalMap[key]) positionalMap[key] = parsed.argv[key]
+          // any new aliases need to be placed in parsedPositionals, so that
+          // the new keys pass strict validation.
+          if (!detailedPositionalParse[key]) {
+            detailedPositionalParse[key] = {
+              value: parsed.argv[key],
+              defaulted: !!parsed.defaulted[key]
```
Comment (mleguen) [inline lib/command.js]: If detailerPositionalParse's list of fields is intended to grow, we should not have to add these fields in 2 places.

I recommand refactoring this part.

File: lib/command.js [hunk 2]
```diff
@@ -348,9 +366,7 @@ module.exports = function command (yargs, usage, validation, globalMiddleware) {
         unparsed.push(value)
       })
     })
-
-    // short-circuit parse.
-    if (!unparsed.length) return
+    if (!unparsed.length) return detailedPositionalParse
```
Comment (mleguen) [inline lib/command.js]: In that case, postProcessPositionals (now runParserOnPositionals) used to return nothing and populatePositionals (now addPositionalsToArgv) used to return positionalMap.

Now, runParserOnPositionals returns `{}` and addPositionalsToArgv returns it too. Isn't something missing? Shouldn't detailedPositionalParse contain at least values from positionalMap before being returned?

**Comment by bcoe** [comment]
@mleguen I agree about concerns of adding complexity, however I also feel we need some way to expose a `detailed` parse from `yargs-parser` (which is an advanced feature @coreyfarrell asked for for `nyc`, and other folks have asked for multiple times throughout the years). If we're going to add this feature, I think we need it to work in `.parse()`, `command()`, and all other locations where a parse occurs.

tldr; I think this ultimately is actually a good replacement to `yargs.parsed`, which only worked in some contexts ... open to other recommendations.

File: test/yargs.js [hunk 3]
```diff
@@ -873,6 +873,127 @@ describe('yargs dsl tests', () => {
...
+          ;(!!parse.defaulted.id).should.equal(false)
+          ;(!!parse.defaulted.foo).should.equal(true)
+          ;(!!parse.defaulted.banana).should.equal(true)
+          parse.aliases.id.should.include('identifier')
+          return done()
+        }, [(argv) => {
+          return {
+            hello: 'world'
+          }
+        }])
```
Comment (coreyfarrell) [inline test/yargs.js]: What is this final callback argument?  Seems like this test could do without it - also with `detailed: true` I'd expect this callback to use `({argv})` argument (parse.argv instead of just argv).

File: test/yargs.js [hunk 4]
```diff
@@ -873,6 +873,127 @@ describe('yargs dsl tests', () => {
...
+            .positional('foo', {
+              default: 'bar'
+            })
+            .alias('id', 'identifier')
+        }, (parse) => {
+          parse.argv.id.should.equal('abc123')
+          parse.argv.identifier.should.equal('abc123')
+          parse.argv.foo.should.equal('bar')
+          parse.argv.banana.should.equal('yellow')
+          parse.argv.hello = 'world'
```
Comment (coreyfarrell) [inline test/yargs.js]: Setting `parse.argv.hello` seems unneeded?

File: test/yargs.js [hunk 5]
```diff
@@ -873,6 +873,127 @@ describe('yargs dsl tests', () => {
...
+            hello: 'world'
+          }
+        }])
+        .parserConfiguration({
+          detailed: true
+        })
+        .option('banana', {
+          default: 'yellow'
+        })
+        .parse('foo abc123 --apple red')
```
Comment (coreyfarrell) [inline test/yargs.js]: As with the yargs-parser tests it would be good if we could use `--apple green` to show that yargs.defaulted.apple is cleared when the default value is explicitly set to the default value by the user.  Also missing a test above to verify parse.defaulted.apple is unset.

File: test/yargs.js [hunk 6]
```diff
@@ -873,6 +873,127 @@ describe('yargs dsl tests', () => {
...
+        .command('foo [id] [foo]', 'foo command', (yargs) => {
+          yargs
+            .option('apple', {
+              default: 'green'
+            })
+            .positional('id', {
+              default: 'blerg'
+            })
+            .positional('foo', {
+              default: 'bar'
```
Comment (coreyfarrell) [inline test/yargs.js]: What is the expected behavior when a default is not configured?  If you remove this `default: 'bar'` would this mean `parse.defaulted.foo == true && parse.argv.foo === undefined`?

File: test/yargs.js [hunk 7]
```diff
@@ -873,6 +873,127 @@ describe('yargs dsl tests', () => {
...
+          detailed: true
+        })
+        .option('banana', {
+          default: 'yellow'
+        })
+        .parse('foo abc123 --apple red')
+      parse.argv.id.should.equal('abc123')
+      parse.argv.identifier.should.equal('abc123')
+      parse.argv.foo.should.equal('bar')
+      parse.argv.banana.should.equal('yellow')
```
Comment (coreyfarrell) [inline test/yargs.js]: Need to check argv.apple and defaulted.apple.

File: test/yargs.js (hunk 3 above)
Comment (bcoe) [reply test/yargs.js]: this is `middleware`, which can be passed as an array as the final argument to a `command`. The idea with middleware is that it might be a third-party plugin that someone has written, so I think the API surface should continue getting `argv` (otherwise someone writing middleware would need to detect a detailed parse, vs., a regular parse).

**Comment by bcoe** [comment]
@coreyfarrell is there a more minimal API surface that would work for your use-case, maybe we could just get away with introducing `.parseDetailed` and ignoring the `.command` use case until someone asks for it.

**Comment by coreyfarrell** [comment]
> ignoring the .command use case until someone asks for it.

For my use case `const {argv, defaults} = yargs.parse(argsArray)` would be adequate, though ignoring the `.command` use case for now would prevent adding it without a breaking change.  Right now setting `detailed: true` has no meaning anywhere, so using it to change the API of both `yargs.parse` and `.command` handlers is not a breaking change.

**Comment by bcoe** [comment]
we have this ancient pull request as well, asking us to expose `defaults`:

https://github.com/yargs/yargs/issues/276

I don't love the idea of adding a `yargs.defaults()` because it feels a bit clunky, why no `yargs.aliases()`, `yargs.etc` ... perhaps a compromise would be a method like:

```js
yargs.isDefaulted(string) // returns true or false depending on whether the optoin was defaulted.
```

☝️ I could imagine adding a handful of options like this.

**Comment by bcoe** [comment]
somewhat similar: https://github.com/yargs/yargs/issues/294, a way to filter out any keys that have aliases.

**Comment by coreyfarrell** [comment]
The API of the `yargs` object is difficult as many of the functions mean "add this thing" rather than "get information about this thing".  So I think `yargs.isDefaulted(string)` makes sense.  FWIW my previous message mentioned `const {argv, defaults} = yargs.parse(argsArray)`, this is incorrect.  It should have said `const {argv, defaulted} = yargs.parse(argsArray)`.  So I think `yargs.defaulted` avoids the confusion that `yargs.defaults` would create, though I say that from the prospective of a native English speaker.  I'm not sure if `defaulted` would be so clear to people who are not native English speakers so maybe `isDefaulted(string)` would be better?  If we go in that direction then should we consistently expose `isDefaulted(string)` instead of `defaulted`?

**Comment by juergba** [comment]
What about the `error` object, which is populated by yargs-parser?
Does it really throw in Yargs? Otherwise it should also be accessible in Yargs to make sense.
In yargs-parser it's populated only and does not throw.

**Comment by mleguen** [comment]
@coreyfarrell As a non-native English speaker, I am used to seeing (and writing) isXxxed functions for this kind of check. So I would recommend using isDefaulted. But that is only ONE non-native english speaker opinion...

**Comment by bcoe** [comment]
@mleguen @coreyfarrell I'm leaning towards the `isDefaulted`, vs., the current state of this branch.

**Comment by bcoe** [comment]
I need to actually sit down and work on the refactor discussed here, i.e., adding a helper method rather than a sweeping change to how parsing works based on a configuration setting.

**Comment by bcoe** [comment]
let's revisit this when someone has time, it would be good to expose the `defaulted` behavior, but I'm not 100% sure of the best implementation just yet.

**Comment by bcoe** [comment]
I think @coreyfarrell suggested something along the lines of a symbol on the `argv` object, which could be used by `yargs.isDefaulted(argv, key)` to check whether a given field was defaulted.
